### DISCUSSION TITLE
Switch from Elasticsearch to OpenSearch

### DISCUSSION THREAD

**Discussion by merkys** [eclipse-rdf4j/rdf4j#3555] [General]
### Problem description

In early 2021 Elasticsearch has changed their license to a [non-free one](https://opensource.org/node/1099). Due to that, users of certain RDF4J functions will not be able to use most recent Elasticsearch versions and enjoy the freedoms of free and open-source licenses.

### Preferred solution

Amazon Web Services have forked the last Apache-2.0-licensed release of Elasticsearch as [OpenSearch](https://github.com/opensearch-project/OpenSearch) and promised to retain the compatibility with Elasticsearch for some time. I believe switching from Elasticsearch to OpenSearch would be a more sustainable solution in order to cater for wider audiences. I found at least [one reference](https://satish1v.medium.com/getting-started-with-opensearch-a113fe6dde0) saying that the replacement of Elasticsearch with OpenSearch is a drop-in one, thus not much work now. Later on these products are likely to diverge.

### Are you interested in contributing a solution yourself?

No

**Comment by barthanssens** [comment]
Could indeed be useful

**Comment by barthanssens** [comment]
Somewhat related to #3396

**Comment by barthanssens** [comment]
@hmottestad / @jeenbroekstra  any thoughts on this ? I guess we could reuse the ES code and do a find/replace for package names/dependency names, and it might be easier to just file new CQs for this Apache-licensed fork instead of having the new ES-license vetted...

**Comment by hmottestad** [comment]
I would prefer not to. The reason Elasticsearch switched their license was because Amazon wouldn't give them a decent share of the profits that Amazon was making by getting Elasticsearch for free and selling it as a cloud service. So Amazons answer was to fork the project and say that they will put man hours into their fork instead of bowing to Elastics demands. Essentially trying to send a message to all open source projects out there that "either you maintain your project free of charge, or we will fork it and tell everyone to ditch your version". Not something I want RDF4J to be associated with.

**Comment by hmottestad** [comment]
Elastic is moving to a new license which is compatible with SSPL. Eclipse has already accepted works-with CQ request with SSPL (Mongodb), so we should be able to continue using Elasticsearch if we can limit ourselves to works-with.

**Comment by abrokenjester** [comment]
I'm not sure I want to come down one side or the other in this debate between OpenSearch and Elasticsearch, not without a good deal more thought and background reading, but I agree with @hmottestad to an extent: a workswith dependency sounds like a good solution. Of course a consequence is that RDF4J users will have to download elasticsearch themselves , but that might prove to be a plus for other reasons: aligning versions in a complex project is tricky with a hard dependency.

**Comment by JervenBolleman** [comment]
Hi, I am blocked on my laptop to run the elastic search tests as I need https://github.com/elastic/elasticsearch/pull/86990 or can't run the elastic search tests anymore. What is the way forward here?

**Reply by hmottestad** [reply]
I think we should migrate to test containers for our tests. That would require everyone to have docker or something equivalent.

**Reply by JervenBolleman** [reply]
I am ok with that. It is on my teams roadmap to move to test containers in the next year. Maybe I can pick up that experience and reuse it for RDF4j.
